File: substitution_cipher.py
import string

def encrypt(message, key):
    # Create a dictionary to map each letter to its corresponding key letter
    alphabet = string.ascii_lowercase
    mapping = dict(zip(alphabet, key))

    # Encrypt the message by replacing each letter with its corresponding key letter
    encrypted = ''
    for letter in message.lower():
        if letter in mapping:
            encrypted += mapping[letter]
        else:
            encrypted += letter

    return encrypted

def decrypt(message, key):
    # Create a dictionary to map each key letter to its corresponding letter
    alphabet = string.ascii_lowercase
    mapping = dict(zip(key, alphabet))

    # Decrypt the message by replacing each key letter with its corresponding letter
    decrypted = ''
    for letter in message.lower():
        if letter in mapping:
            decrypted += mapping[letter]
        else:
            decrypted += letter

    return decrypted

# Example usage
message = 'hello world'
key = 'qwertyuiopasdfghjklzxcvbnm'
encrypted = encrypt(message, key)
decrypted = decrypt(encrypted, key)

print('Original message:', message)
print('Encrypted message:', encrypted)
print('Decrypted message:', decrypted)

File: substitution_cipher.cpp
#include <iostream>
#include <string>

using namespace std;

string encrypt(string message, string key) {
    string ciphertext = "";
    for (int i = 0; i < message.length(); i++) {
        if (message[i] >= 'a' && message[i] <= 'z') {
            ciphertext += key[message[i] - 'a'];
        } else if (message[i] >= 'A' && message[i] <= 'Z') {
            ciphertext += toupper(key[message[i] - 'A']);
        } else {
            ciphertext += message[i];
        }
    }
    return ciphertext;
}

string decrypt(string ciphertext, string key) {
    string message = "";
    for (int i = 0; i < ciphertext.length(); i++) {
        if (ciphertext[i] >= 'a' && ciphertext[i] <= 'z') {
            message += 'a' + key.find(ciphertext[i]);
        } else if (ciphertext[i] >= 'A' && ciphertext[i] <= 'Z') {
            message += 'A' + key.find(tolower(ciphertext[i]));
        } else {
            message += ciphertext[i];
        }
    }
    return message;
}

int main() {
    string message = "hello world";
    string key = "qwertyuiopasdfghjklzxcvbnm";
    string ciphertext = encrypt(message, key);
    string decrypted_message = decrypt(ciphertext, key);
    cout << "Original message: " << message << endl;
    cout << "Encrypted message: " << ciphertext << endl;
    cout << "Decrypted message: " << decrypted_message << endl;
    return 0;
}
